src/gui/guiHyperText.cpp

```diff
@@ -0,0 +1,1069 @@
+				m_element->dim.Height = height;
+		}
+		endElement();
+
+	} else if (name == "tag") {
+		// Required attributes
+		if (!attrs.count("name"))
+			return 0;
+
+		StyleList tagstyle;
+		parseStyles(attrs, tagstyle);
+
+		if (attrs["paragraph"] == "true")
+			m_paragraphtags[attrs["name"]] = tagstyle;
+		else
+			m_elementtags[attrs["name"]] = tagstyle;
+
+	} else if (name == "action") {
+		if (end)
+			closeTag(name);
+		else {
+			if (!attrs.count("name"))
+				return 0;
+			openTag(name, attrs)->style = m_elementtags["action"];
+		}
+
+	} else if (m_elementtags.count(name)) {
+		if (end)
+			closeTag(name);
+		else
+			openTag(name, attrs)->style = m_elementtags[name];
+		endElement();
+
+	} else if (m_paragraphtags.count(name)) {
+		if (end)
+			closeTag(name);
+		else
+			openTag(name, attrs)->style = m_paragraphtags[name];
+		endParagraph();
+
+	} else
+		return 0; // Unknown tag
+
+	// Update styles accordingly
+	m_style.clear();
+ 	for (auto tag = m_active_tags.crbegin(); tag != m_active_tags.crend(); ++tag)
+		for (auto const &prop : (*tag)->style)
+			m_style[prop.first] = prop.second;
+
+	return cursor;
+}
+
+// -----------------------------------------------------------------------------
+// Text Drawer
+
+TextDrawer::TextDrawer(
+	const wchar_t* text,
+	Client* client,
+	gui::IGUIEnvironment* environment,
+	ISimpleTextureSource *tsrc) :
+	m_text(text), m_client(client), m_environment(environment)
+{
+	// Size all elements
+	for (auto &p : m_text.m_paragraphs) {
+		for (auto &e : p.elements) {
+			if (e.type == ParsedText::ELEMENT_SEPARATOR ||
+					e.type == ParsedText::ELEMENT_TEXT) {
+				if (e.font) {
+					e.dim.Width = e.font->getDimension(e.text.c_str()).Width;
+					e.dim.Height = e.font->getDimension(L"Yy").Height;
+				} else
+					e.dim = {0,0};
+
+			} else if (e.type == ParsedText::ELEMENT_IMAGE ||
+					e.type == ParsedText::ELEMENT_ITEM) {
+				// Dont resize already sized items (sized by another mechanism)
+				if (e.dim.Height == 0 || e.dim.Width == 0) {
+					// Default image and item size
+					core::dimension2d<u32> dim(80, 80);
+
+					if (e.type == ParsedText::ELEMENT_IMAGE) {
+						video::ITexture *texture =
+							m_client->getTextureSource()->
+								getTexture(strwtostr(e.text));
+						if (texture)
+							dim = texture->getOriginalSize();
+					}
+
+					if (e.dim.Height == 0)
+						if (e.dim.Width == 0)
+							e.dim = dim;
+						else
+							e.dim.Height = dim.Height * e.dim.Width / dim.Width;
+					else
+						e.dim.Width = dim.Width * e.dim.Height / dim.Height;
+				}
+			}
+		}
+	}
+}
+
+// Get element at given coordinates. Coordinates are inner coordinates (starting
+// at 0,0).
+ParsedText::Element* TextDrawer::getElementAt(const core::position2d<s32> &pos)
+{
+	for (auto &p : m_text.m_paragraphs) {
+		for (auto &el : p.elements) {
+			core::rect<s32> rect(el.pos, el.dim);
+			if (rect.isPointInside(pos))
+				return &el;
+		}
+	}
+	return 0;
+}
+
+void TextDrawer::place(s32 width)
```

Comment:
Would you please be so nice to include a quick summary about what this function in general does?

Reply:
This function places all elements according to given width. Elements have been previously sized by constructor and will be later drawed by draw. It may be called each time width changes and resulting height is stored in m_height. See GUIHyperText constructor, it uses it once to test if text fits in window and eventually another time if width is reduced because of scrollbar added.

(this comment will be added to code)